README.md
# Introduction

molUP is a VMD extension that provides a simple manner for loading and saving Gaussian files, and analyze related results. molUP provides a graphical interface for VMD where the users can load and save chemical structures in the Gaussian file formats. This extension includes a set of tools to set up any calculation supported by Gaussian, including ONIOM; analyze energies through interactive plots; animate vibrational frequencies; draw the vectors associated with those frequencies; modify bonds, angles, and dihedrals; and collect bibliographic information on the employed methods.

Main Features:
- Load Gaussian Input (.com) and Output files (.log) on VMD;
- Assign ONIOM layers (High-, Medium-, and Low-level);
- Select free and fixed atoms during geometry optimization (Freeze status);
- Analyze and edit atomic charges;
- Setup new Gaussian calculation using the Gaussian input section;
- Load scan across reactional coordinates and plot the structure's energies;
- Load Gaussian frequency calculations, animate the vibrational modes and show respective vectors;
- Show quick useful representations such as the protein, the different ONIOM layers, and the fixed atoms;
- Structural Manipulation: adjust the bond length, angle amplitude, and dihedral angle torsion;
- Provide a complete bibliographic list of references according to the type of methods and functionals that were employed in the calculation;
- Add atoms and molecular fragments NEW;
- Remove atoms NEW;
- Among others.

![Logo](Screenshots/image1.gif)

![Adjusting a dihedral angles using molUP](Screenshots/image2.gif)

![Energies](Screenshots/image3.gif)

![Input section](Screenshots/image4.gif)

![ONIOM layers](Screenshots/image5.gif)

![Loading a file](Screenshots/image6.gif)

## Minimum Requirements

Operating System: macOS, Linux, and Windows

Visual Molecular Dynamics (VMD) 1.9.3 or later

## Installation

**Automatic Installation**

You could easily install molUP through [vmdStore](https://github.com/BioSIM-Research-Group/vmdStore).


**Manual Installation**

Alternatively, you could install molUP manually:
1. Download or Clone the repository.
2. Copy the molUP directory to a location on you computer. (Get that location!)
3. Copy the text inside "install.txt" file and paste on your .vmdrc file (macOS and Linux ) or vmd.rc file (Windows).
4. Replace the string "$::vmdStorePath/plugins/molUP" by your installation location (Step 2).
5. Save the .vmdrc or vmd.rc file.
6. Restart VMD.


**(Optional) Run molUP from bash**

molUP supports now the loading of Gaussian files directly from your command line (available for macOS and Linux).

Tutorial:
1. Add the following line to your .bashrc (Linux) or .bash_profile (macOS) file. (This file is located in the HOME directory) You have to edit the "path of the bashScript.tcl file" field by the complete path of this file that is located in the molUP installation directory: "molUP/lib/bashScript.tcl"

alias molUP='vmd -e "path of the bashScript.tcl file" -args'

2. Call molUP from the command line:

Open a Gaussian file using the standard procedure:

molUP "Gaussian file"



Open a Gaussian output file reading all available structures:

molUP "Gaussian file" -all


## Contacts
If you have any suggestion of new features, please contact us: hfernandes@med.up.pt

## Citation
S. Fernandes, H., M.J. Ramos, and N. M. F. S. A. Cerqueira, molUP: A VMD plugin to handle QM and ONIOM calculations using the gaussian software. Journal of Computational Chemistry

## DOI
http://doi.org/10.1002/jcc.25189


molUP/changelog.txt
######### molUP ##############
Last version: 1.7.1
##############################
Version 1.7.1 (14 JAN 2023)
- [New feature] Implemented a method to load AMBER (PRMTOP file) parameters from the terminal (Thanks to NithinChintala).

Version 1.7.0 (03 MAY 2021)
- [New feature] Now, molUP saves ORCA input files with pointcharges file according to the ONIOM Electrostatic Embedding Scheme: ONIOM=ScaleCharge=555500 (Default values of Gaussian Software).

Version 1.6.8 (15 MAR 2021)
- [Bug fixed] An issue prevented some Gaussian input files, created by other software, to properly load in molUP.

Version 1.6.7 (9 MAR 2021)
- [Bug fixed] An issue prevented the animation of vibrational frequencies in some Gaussian output files.

Version 1.6.6 (21 JAN 2021)
- [Bug fixed] An issue prevented some Gaussian output files to load. This only affected Gaussian calculations restarted from checkpoint files.

Version 1.6.5 (29 JUN 2020)
- [Bug fixed] Some small vectors were not being displayed. 

Version 1.6.4 (25 JUN 2020)
- [Bug fixed] Wrong element on atomtypes starting with numbers.

Version 1.6.3 (30 APR 2020)
- [Improvement] Now, molUP requires GNU-sed and GNU-grep when running on macOS. gsed and ggrep commands should be available on macOS. To install them, please install HomeBrew: /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)" and then: brew install grep gnu-sed
- [Improvement] molUP is now able to load Gaussian files whose coordinates are structured in the following format: " <ELEMENT>,<CHARGE>,<XXX>,<YYY>,<ZZZ>"

Version 1.6.2 (8 JAN 2020)
- [New feature] A simple Tk console command was added to easily rename residues' names (Resname). Run "::molUP::changeResname VMD_SELECTION NEW_RESNAME", e.g.: "::molUP::changeResname "resid 53" VAL"

Version 1.6.1 (7 JAN 2020)
- [Improvement] Additional molecular groups were added to the "Add Molecule/Fragment" tool.

Version 1.6.0 (28 SEP 2019)
- [Improvement] molUP is now able to load Gaussian files whose coordinates are structured in two other formats.
- [New feature] molUP has a new interactive selector to navigate though the frames.

Version 1.5.4 (15 FEV 2019)
- [New feature] Atom selections can be used to choose which atoms must be deleted when using "Delete Atoms" tool.
- [Bug fixed] A minor bug related to an error message shown during the VMD initialization was fixed.

Version 1.5.3 (30 OUT 2018)
- [Bug fixed] An issue that hide the button "Add" on the "Add Molecule/Fragment" window of the "Add Atoms" tool. 

Version 1.5.2 (18 OUT 2018)
- [New feature] A button was added to allow the import of new calculated charges from Gaussian output files.
- [New feature] Atoms can now be picked to assign them to an ONIOM layer or to change its freezing state during optimization calculations.
- [Minor Change] The charge and multiplicity are now loaded from the Gaussian file instead of being calculated ad the beginning.
- [Bug fixed] Fixed the issue that prevented the VMD citation appears in the references list.

Version 1.5.1 (6 SEP 2018)
- [Minor Changes] Now all the internal packages have their own versions.

Version 1.5.0 (1 AUG 2018)
- [New feature] molUP now supports saving ORCA Input files.

Version 1.4.12 (31 JUL 2018)
- [Bug fixed] Now, the link atoms are placed at the correct distance when custom selections are used to save new Gaussian Input Files.

Version 1.4.11 (19 JUN 2018)
- [Bug fixed] Elements with two characters were incorrectly saved when new Gaussian input files were generated.

Version 1.4.10 (18 JUN 2018)
- [Bug fixed] Fixed an issue that misunderstanding the elements when new Gaussian input files are saved.

Version 1.4.9 (21 MAY 2018)
- [Bug fixed] Fixed an issue that prevents the loading of multiple files using molUP.

Version 1.4.8 (17 MAY 2018)
- [Bug fixed] Fixed an issue that prevents the loading of PRMTOP files that could present an unusual format.

Version 1.4.7 (16 MAY 2018)
- [Improvement] Significant improvement of the procedure that imports the PRMTOP file.

Version 1.4.6 (15 MAY 2018)
- [New feature] A "-all" flag was added to the molUP bash command. This new feature is allow the loading of all structures of a any Gaussian Output file.
- [Bug fixed] Fixed an issue that precluded the conversion of certain AMBER parameters from PRMTOP files.

Version 1.4.5 (10 MAY 2018)
- [Bug fixed] Fixed an issue that does not allow the correct export of the energy data from energy profiles.

Version 1.4.4 (9 ABR 2018)
- [Bug fixed] Fixed an issue that originated an error when modification tools were used after the addition or deletion of atoms.
- [Bug fixed] Fixed an issue that generated random behavior when modification tools were used.

Version 1.4.3 (9 ABR 2018)
- [Bug fixed] Fixed an issue that changes the connectivity when a new molecule is loaded.

Version 1.4.2 (5 ABR 2018)
- [Bug fixed] Fixed an issue that prevents the visualization of frequencies.

Version 1.4.1 (3 ABR 2018)
- [Bug fixed] Fixed an issue that changed the view during the process to add and/or remove atoms.
- [Bug fixed] Fixed an issue that displayed an error message when the user clicked on "Delete" and no atom had been selected.

Version 1.4.0 (3 ABR 2018)
- [New feature] Now you can delete and add atoms using molUP.
- [Improvement] The footer toolbar includes a shortcut button to enable the Add/Remove Bonds mouse mode, and two icons to use the new tools that enable the addition and deletion of atoms.

Version 1.3.10 (9 MAR 2018)
- [Bug fixed] Fixed a bug that prevented the selection of the last three frequencies.

Version 1.3.9 (7 MAR 2018)
- [Bugs fixed] Minor bugs fixed
- [New feature] Add custom selections in the modification (bonds, angles, and dihedral angles) tools

Version 1.3.8 (5 MAR 2018)
- [Bug fixed] Fixed an issue that gave an error collection energies from the last structure of certain Gaussian output files.

Version 1.3.7 (2 MAR 2018)
- [New feature] Load multiple structures at once.


molUP/molUP.tcl
package provide molUP 1.7.1

#### INIT ############################################################
namespace eval molUP:: {
	namespace export molUP
	
		#### Load Packages				
		package require guiMolUP 								1.6.0
		package require guiBondModif							1.5.1
		package require guiAngleModif							1.5.1
		package require guiDihedModif							1.5.1
		package require guiOpenFile								1.5.1
		package require guiSaveFile								1.5.1
		package require guiError								1.5.1
		package require guiChargeMulti							1.5.1
		package require guiCalcSetup							1.5.1
		package require guiCredits								1.5.2
		package require guiChangelog							1.5.1
		package require guiInfo 								1.5.1
		package require guiMethodology							1.5.1
		package require guiBADparam								1.5.1
		package require guiModRedundant							1.5.1
		package require guiOpenMultiFile 						1.5.1
		package require guiDeleteAtoms							1.5.4
		package require guiAddAtoms								1.5.1
		package require guiAddMolecule 							1.5.3
		
		package require inputFile 								1.6.8
		package require timeControl								1.5.1
		package require quit									1.5.1
		package require loadGaussianInputFile					1.6.3
		package require loadGaussianOutputFile					1.6.6
		package	require editStructure							1.5.2
		package require modify									1.5.1
		package require saveFiles								1.7.0
		package require readFreq								1.6.7
		package require energy									1.5.1
		package require plot									1.5.1
		package require addDeleteAtoms							1.5.4
		package require publication								1.5.2
		package require loadPrmtop								1.5.1
		package require loadPrmtopTerminal  					1.7.1
		package require updateChargesFromFile					1.5.2
		package require changeStructureData						1.6.2


		package require Tk
		package require tablelist
		package require topotools
		package require balloon
		
		# Theme
		package require gaussianTheme							1.5.1

		#### Program Variables
		## General
		variable version	    	"1.7.1"

		#GUI
        variable topGui         	".molUP"
		variable bondModif         	".molUP.bondModif"
		variable angleModif        	".molUP.angleModif"
		variable dihedModif        	".molUP.dihedModif"
		variable openFile        	".molUP.openFile"
		variable openMultiFile		".molUP.openMultiFile"
		variable saveFile        	".molUP.saveFile"
		variable error	        	".molUP.error"
		variable chargeMulti	    ".molUP.chargeMulti"
		variable calcSetup			".molUP.calcSetup"
		variable credits			".molUP.credits"
		variable changelog			".molUP.changelog"
		variable citeMolUP			".molUP.publication"
		variable citations			".molUP.citations"
		variable info				".molUP.info"
		variable methodology		".molUP.methodology"
		variable saveKeywordsInput	".molUP.saveKeywordsInput"
		variable modRedundant		".molUP.modRedundant"
		variable deleteAtoms		".molUP.deleteAtoms"
		variable addAtoms			".molUP.addAtoms"
		variable periodicTable		".molUP.addAtoms.periodicTable"
		variable addMolecule		".molUP.addAtoms.addMolecule"
		variable updateCharges		".molUP.updateCharges"

		global path
		variable path 				"/"
		variable fileName			""
		variable fileExtension		""
		variable title 				"molUP is awesome"
		variable actualTitle		"molUP is awesome"
		variable chargesMultip		""
		variable keywordsCalc		"%mem=7000MB\n%NProc=4\n%chk=name.chk\n# "
		variable structureGaussian	""
		variable loadingProgress	"0.0"
		variable atomicSymbolList 		
		variable gaussianAtomTypeList	
		variable pdbAtomTypeList	 	
		variable resnameList		 	
		variable residList		 		
		variable chargeList 			
		variable freezeList				
		variable xxList					
		variable yyList					
		variable zzList					
		variable atomDesigList			
		variable linkAtomList			
		variable linkAtomNumbList	
		variable linkAtomValueList
		variable numberAtoms
		variable numberStructures
		variable actualTime
		variable temporaryPDBFile
		variable temporaryXYZFile
		variable time0
		variable time1
		variable loadMode						""
		variable selectionModificationType		""
		variable selectionModificationValueOniom		""
		variable selectionModificationValueFreeze		""
		variable atomSelectionONIOM				"none"
		variable atomSelectionFreeze			"none"
		variable HLrep		"0"
		variable MLrep		"0"
		variable LLrep		"0"
		variable unfreezeRep	"0"
		variable showNegChargedResidues	"0"
		variable showPosChargedResidues	"0"
		variable freezeRep		"0"
		variable allRep			"1"
		variable proteinRep		"0"
		variable nonproteinRep	"0"
		variable waterRep		"0"
		variable pickedAtoms	{}
		variable atom1BondSel	"none"
		variable atom2BondSel	"none"
		variable atom1BondOpt	"Fixed Atom"
		variable atom2BondOpt	"Move Atom"
		variable atom1AngleSel	"none"
		variable atom2AngleSel	"none"
		variable atom3AngleSel	"none"
		variable atom1AngleOpt	"Fixed Atom"
		variable atom2AngleOpt	""
		variable atom3AngleOpt	"Move Atom"
		variable atom1DihedSel	"none"
		variable atom2DihedSel	"none"
		variable atom3DihedSel	"none"
		variable atom4DihedSel	"none"
		variable atom1DihedOpt	"Fixed Atom"
		variable atom2DihedOpt	""
		variable atom3DihedOpt	""
		variable atom4DihedOpt	"Move Atom"
		variable customSelection1		""
		variable customSelection2		""
		variable BondDistance			"0.01"
		variable initialBondDistance	"0.01"
		variable AngleValue				"0.00"
		variable initialAngleValue		"0.00"
		variable DihedValue				"0.00"
		variable initialDihedValue		"0.00"
		variable initialSelection {}
		variable initialSelectionX {}
		variable initialSelectionY {}
		variable initialSelectionZ {}
		variable pos1
		variable pos2
		variable pos3
		variable pos4
		variable normvec
		variable linkAtomsList		{}
		variable linkAtomsListIndex	{}
		variable saveOptions	"Gaussian Input File (.com)"
		variable connectivityInputFile ""
		variable connectivity ""
		variable parameters ""
		variable openNewFileMode "YES"
		variable tmpFolder ""
		variable structureReadyToLoad {}
		variable structureReadyToLoadCharges {}
		variable structureReadyToLoadLayer {}
		variable structureReadyToLoadFreeze {}
		variable moleculeInfo {}
		variable majorHeight
		variable freqVectorColor 		"1 red"
		variable colorList {{0 blue} {1 red} {2 gray} {3 orange} {4 yellow} {5 tan} {6 silver} {7 green} {8 white} {9 pink} {10 cyan} {11 purple} {12 lime} {13 mauve} {14 ochre} {15 iceblue} {16 black} {17 yellow2} {18 yellow3} {19 green2} {20 green3} {21 cyan2} {22 cyan3} {23 blue2} {24 blue3} {25 violet} {26 violet2} {27 magenta} {28 magenta2} {29 red2} {30 red3} {31 orange1} {32 orange3}}
		variable pathsFreq {}
		variable nFreqToRead 5
		variable vectorThreshold 0
		variable saveAdvancedOptions "All"
		variable atomSelectionSave "all"
		variable addAtomElement	"H"
		variable moveAtomPosRectInitial	{}
		variable atomSelectionDeleteAtoms ""


		variable chargeAll	0
		variable chargeHL	0
		variable chargeML	0
		variable chargeLL	0
		variable multiplicityValue		1
		variable multiplicityValue1		1
		variable multiplicityValue2		1

		variable normalTermination	"NO"
		variable stopLine 1


		variable graphicsID ""
    	variable pickedAtomsBAD {}

		## grep, sed, head, cut, and tail
		if {[string first "Windows" $::tcl_platform(os)] != -1} {
			variable sed "$molUPpath/windowsDependencies/sed.exe"
			variable head "$molUPpath/windowsDependencies/head.exe"
			variable cut "$molUPpath/windowsDependencies/cut.exe"
			variable tail "$molUPpath/windowsDependencies/tail.exe"
			variable grep "$molUPpath/windowsDependencies/grep.exe"
		} elseif {[string first "Darwin" $::tcl_platform(os)] != -1} {
			catch {exec ggrep} debug
			if {[string match "*no such file or directory*" $debug] == 1 } {
				puts "ERROR-ERROR-ERROR - molUP - ERROR-ERROR-ERROR\nPlease, install GNU-grep. In Terminal, run the following commands:\n/bin/bash -c \"\$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)\"\nbrew install grep\n\nIf you have already Homebrew installed on your machine, the first command is not required. Simply run:\nbrew install grep"
			}
			catch {exec gsed} debug
			if {[string match "*no such file or directory*" $debug] == 1 } {
				puts "ERROR-ERROR-ERROR - molUP - ERROR-ERROR-ERROR\nPlease, install GNU-grep. In Terminal, run the following commands:\n/bin/bash -c \"\$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install.sh)\"\nbrew install gnu-sed\n\nIf you have already Homebrew installed on your machine, the first command is not required. Simply run:\nbrew install gnu-sed"
			}
			variable sed "gsed"
			variable head "head"
			variable cut "cut"
			variable tail "tail"
			variable grep "ggrep"
		} else {
			variable sed "sed"
			variable head "head"
			variable cut "cut"
			variable tail "tail"
			variable grep "grep"
		}
		
}

proc molUP::start {} {
	encoding system utf-8
	if {[winfo exists $::molUP::topGui]} {
		wm deiconify $::molUP::topGui
		return $::molUP::topGui
		}

	### Reload keywords to color tag text
	molUP::readKeywordsTags

	### Open the GUI
	molUP::buildGui
	update
	return $::molUP::topGui
}
